### PR TITLE
parity for Sink, for preMaterialize operation

### PR DESCRIPTION
**Pull request by ktoso**
Resolves https://github.com/akka/akka/issues/24656 

Provides feature parity between preMaterialize on a Sink and a Source.

TODO: 

- [x] add tests
- [x] add java api

### CODE REVIEW

**Review by 2m — APPROVED** [review]

File: akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala [hunk 1]
```diff
@@ -321,6 +322,17 @@ final class Sink[-In, +Mat](delegate: scaladsl.Sink[In, Mat]) extends Graph[Sink
   def mapMaterializedValue[Mat2](f: function.Function[Mat, Mat2]): Sink[In, Mat2] =
     new Sink(delegate.mapMaterializedValue(f.apply _))
 
+  /**
+   * Materializes this Sink, immediately returning (1) its materialized value, and (2) a new Sink
+   * that can be consume elements 'into' the pre-materialized one.
+   *
+   * Useful for when you need a materialized value of a Sink when handing it out to someone to materialize it for you.
+   */
+  def preMaterialize(materializer: Materializer): japi.Pair[Mat @uncheckedVariance, scaladsl.Sink[In @uncheckedVariance, NotUsed]] = {
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: scaladsl.Sink in javadsl

Comment (patriknw) [reply akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: are these  uncheckedVariance related to https://github.com/akka/akka/pull/24575 ?
add java api compile test?

**Review by patriknw — DISMISSED** [review]

File: akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala (hunk 1 above)
Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: Whoops, thanks

**Comment by ktoso** [comment]
Not entirely sure if #24575 resolves the 

```
[error] /Users/ktoso/code/akka/akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala:331:7: contravariant type In occurs in invariant position in type (materializer: akka.stream.Materializer)akka.japi.Pair[Mat @scala.annotation.unchecked.uncheckedVariance,akka.stream.javadsl.Sink[In,akka.NotUsed]] of method preMaterialize
[error]   def preMaterialize(materializer: Materializer): japi.Pair[Mat @uncheckedVariance, Sink[In, NotUsed]] = {
```

that I had here?

**Review by ktoso — COMMENTED** [review]
Added docs and java test, should be now ready to merge

**Comment by ktoso** [comment]
Failure was https://github.com/akka/akka/issues/24672 it indeed is a bug it seems, we'l have to fix it

**Pull request merged**